lib/src/main/kotlin/org/didcommx/didcomm/protocols/routing/Forward.kt

```diff
@@ -0,0 +1,231 @@
+const val PROFILE_DIDCOMM_AIP2_ENV_RFC587 = "didcomm/aip2;env=rfc587"
+const val PROFILE_DIDCOMM_V2 = "didcomm/v2"
+
+internal fun findDIDCommService(
+    didDocResolver: DIDDocResolver,
+    to: String,
+    serviceId: String? = null
+): DIDCommService? {
+
+    val toDid = getDid(to)
+    val didDoc = didDocResolver.resolve(toDid).orElseThrow { throw DIDDocNotResolvedException(toDid) }
+
+    if (serviceId != null) {
+        val didService = didDoc.findDIDCommService(serviceId)
+
+        if (PROFILE_DIDCOMM_V2 !in didService.accept) {
+            throw DIDCommServiceException(
+                toDid, "service '$serviceId' does not accept didcomm/v2 profile"
+            )
+        }
+        return didService
+    } else {
+        // Find the first service accepting `didcomm/v2` profile because the spec states:
+        // > Entries SHOULD be specified in order of receiver preference,
+        // > but any endpoint MAY be selected by the sender, typically
+        // > by protocol availability or preference.
+        // https://identity.foundation/didcomm-messaging/spec/#multiple-endpoints
+        return try {
+            didDoc.didCommServices.find { PROFILE_DIDCOMM_V2 in it.accept }
+        } catch (e: DIDDocException) {
+            null
+        }
+    }
+}
+
+internal fun resolveDIDCommServicesChain(
+    didDocResolver: DIDDocResolver,
+    to: String,
+    serviceId: String? = null,
+    didRecursion: Boolean = false
+): List<DIDCommService> {
+
+    val toDidService = findDIDCommService(didDocResolver, to, serviceId) ?: return listOf()
+
+    val res = mutableListOf<DIDCommService>()
+    var serviceUri = toDidService.serviceEndpoint
+
+    res.add(0, toDidService)
+
+    // alternative endpoints
+    while (isDIDOrDidUrl(serviceUri)) {
+        val mediatorDid = serviceUri
+
+        if (res.size > 1) {
+            // TODO cover possible case of alternative endpoints in mediator's
+            //      DID Doc services (it SHOULD NOT be as per spec but ...)
+            val errMsg = (
+                "mediator '${res.last().serviceEndpoint}' defines alternative" +
+                    " endpoint '$serviceUri' recursively"
+                )
+
+            if (didRecursion) {
+                throw NotImplementedError(errMsg)
+            } else {
+                throw DIDCommServiceException(res.last().serviceEndpoint, errMsg)
+            }
+        }
+
+        // TODO check not only first item in mediator services list
+        //      (e.g. first one may use alternative endpoint but second - URI)
+
+        // resolve until final URI is reached
+        val mediatorDidService = findDIDCommService(didDocResolver, mediatorDid)
+            ?: throw DIDCommServiceException(
+                mediatorDid, "mediator '$mediatorDid' service doc not found"
+            )
+
+        serviceUri = mediatorDidService.serviceEndpoint
+        res.add(0, mediatorDidService)
+    }
+
+    return res
+}
+
+fun wrapInForward(
```

Comment:
If we consider wrapInForward and unpackForward as public API (though advanced one), I believe they should not expose internals such as SenderKeySelector, and have only secrets and did resolvers as parameters (as methods in DIDComm class). Following the same pattern as for DIDComm, probably we could create DIDCommForward class with two public methods wrapInForward() and unpackForward().